### DISCUSSION TITLE
Pasting current date on global hotkey only if specific apps have the focus

### DISCUSSION THREAD

**Discussion by Gitoffthelawn** [hluk/CopyQ#1647] [Q&A] [answered]
I would like to paste the current date on global hotkey only if specific apps have the focus (for example, `vim` and `Notepad++`).  Because these are text editors, we know the window name will end with one of those two strings, but may also contain characters before it.

How can this be accomplished?

**Comment by hluk** [comment, marked as answer]
You can use [currentWindowTitle()](https://copyq.readthedocs.io/en/latest/scripting-api.html#String%20currentWindowTitle).

```js
if (/Konsole$/.test(currentWindowTitle())) {
    // Do something if a "Konsole" window has focus.
}
```

**Reply by Gitoffthelawn** [reply]
Thank you!  I was so close, but was trying to use boolean expressions instead of `test`.

**Reply by Gitoffthelawn** [reply]
@hluk Given the tip you provided, I've gotten closer to what I'm trying to do.

Here's what I have so far:

```
copyq:
if (/vim$/.test(currentWindowTitle())) {
   var dateTime = dateString('yyyy/MM/dd hh:mm')
   copy(dateTime)
   copySelection(dateTime)
   paste()
}
```

I have a few issues using this:
 1. It changes the clipboard content.  I don't want the clipboard to change at all.
 2. It shows a notification with 'Text copied' and the timestamp (I'm using system notifications, if that matters).
 3. It's surprisingly slow to respond.  Nothing horrible, but it takes maybe 400-500ms for the datestamp to get inserted.
 4. I want this to be ignored on all other windows.  As is, it eats the shortcut key (F5) when pressed!

Can you (or others) help?

**Reply by hluk** [reply]
> 1. It changes the clipboard content. I don't want the clipboard to change at all.

CopyQ needs to change the clipboard to be able to paste the date. You could restore the previous clipboard with (following is just for text):

```js
var previousClipboardText = clipboard()
copy(dataTime)
paste()
sleep(500)  // wait 500 ms after sending paste shortcut, it might be needed depending on the target application
copy(previousClipboardText)
```

> 2. It shows a notification with 'Text copied' and the timestamp (I'm using system notifications, if that matters).

Notifications for CopyQ owned clipboard can be avoided by adding following command:

```ini
[Command]
Command="
    var showDataNotification_ = showDataNotification
    showDataNotification = function() {
        if (!dataFormats().includes(mimeOwner))
            showDataNotification_()
    }"
Icon=\xf1f6
IsScript=true
Name=Avoid Notifications for Own Clipboard
```

> 3. It's surprisingly slow to respond. Nothing horrible, but it takes maybe 400-500ms for the datestamp to get inserted.

On Windows there is `window_wait_after_raised_ms` option which indicates how long to wait for a target window to be raised (by default 150ms on Windows).

You can try setting it to zero (e.g. from Action dialog - F5 shortcut from the main window):

```
copyq config window_wait_after_raised_ms 0
```

> 4. I want this to be ignored on all other windows. As is, it eats the shortcut key (F5) when pressed!

That's not possible in CopyQ. A global shortcut is either active all the time or disabled completely.

**Reply by Gitoffthelawn** [reply]
@hluk Thank you so much for all your help. 👍🏾 

>> 4.  I want this to be ignored on all other windows. As is, it eats the shortcut key (F5) when pressed!

> That's not possible in CopyQ. A global shortcut is either active all the time or disabled completely.

Hmmm.... that's not good! I'm trying to think of possible solutions.  Here are 2 ideas:

1. Allow the user to specify whether a global hotkey should pass-through when pressed.  In other words, CopyQ performs the commands assigned to the hotkey, and then sends the global hotkey just like it was pressed.  Actually, is this already possible somehow?

or

2. In CopyQ's  `Commands` dialog box, there is the `Actions` tab.  This tab has exactly the opposite functionality that I'm looking for, so I wonder if the inverse can be implemented.  It allows only using text copied *from* specific windows; I want to past text only *to* specific windows.  I'm thinking that perhaps using the `Actions` tab, the global hotkey could only be honored on certain windows.  Just a thought.

**Reply by hluk** [reply]
CopyQ can send only shortcuts to paste or copy and that should be an implementation detail. Generally, CopyQ does **not support passing key presses** to applications and it is not the purpose of the clipboard manager. I'm not even sure if pass-through would be possible to implement on supported platforms. I'm certainly not willing to maintain the additional platform-dependent code.

> In CopyQ's Commands dialog box, there is the Actions tab.

I guess you mean "Advanced" tab.

It would be nice for CopyQ to able to set up global shortcuts depending on currently active window changes. There is currently no reliable way to monitor these changes (or if current window title changes) ... but maybe I can improve the code a bit so this is somewhat possible.
